crates/iota-transaction-builder/src/builder.rs

```diff
@@ -0,0 +1,1066 @@
+            module: Identifier::new(module)
+                .unwrap_or_else(|_| panic!("invalid identifier: {module}")),
+            function: Identifier::new(function)
+                .unwrap_or_else(|_| panic!("invalid identifier: {function}")),
+            args: None,
+            generics: Ok(PhantomData),
+            ptb,
+        }
+    }
+
+    /// Set the call params. Optional.
+    pub fn params(
+        self,
+        params: impl IntoIterator<Item = ObjectReference>,
+    ) -> MoveCallCommandBuilder<'a, (), G, Vec<Argument>> {
+        MoveCallCommandBuilder {
+            package: self.package,
+            module: self.module,
+            function: self.function,
+            args: Some(
+                params
+                    .into_iter()
+                    .map(|o| {
+                        self.ptb.input(
+                            InputKind::Input(iota_types::Input::ImmutableOrOwned(o)),
+                            false,
+                        )
+                    })
+                    .collect::<Vec<_>>(),
+            ),
+            generics: self.generics,
+            ptb: self.ptb,
+        }
+    }
+
+    /// Set the generic type arguments. Optional.
+    pub fn generics<U: MoveTypes>(self) -> MoveCallCommandBuilder<'a, (), U, Vec<Argument>> {
+        MoveCallCommandBuilder {
+            package: self.package,
+            module: self.module,
+            function: self.function,
+            args: self.args,
+            generics: Ok(PhantomData),
+            ptb: self.ptb,
+        }
+    }
+
+    /// Set the type arguments manually. Optional.
+    pub fn type_tags(
+        self,
+        tags: impl IntoIterator<Item = TypeTag>,
+    ) -> MoveCallCommandBuilder<'a, (), (), Vec<Argument>> {
+        MoveCallCommandBuilder {
+            package: self.package,
+            module: self.module,
+            function: self.function,
+            args: self.args,
+            generics: Err(tags.into_iter().collect()),
+            ptb: self.ptb,
+        }
+    }
+
+    /// Finish the move call and return the PTB.
+    pub fn end(self) -> &'a mut TransactionBuilder<()> {
+        let args = self.args.unwrap_or_default();
+
+        self.ptb.command(Command::MoveCall(MoveCall {
+            package: self.package,
+            module: self.module,
+            function: self.function,
+            type_arguments: match self.generics {
+                Ok(_) => G::type_tags(),
+                Err(t) => t,
+            },
+            arguments: args,
+        }));
+
+        self.ptb
+    }
+
+    /// Finish the move call by naming the output and return the PTB.
+    pub fn result(self, name: impl NamedCommands) -> &'a mut TransactionBuilder<()> {
+        let ptb = self.end();
+
+        name.push_named_commands(ptb);
+
+        ptb
+    }
+}
+
+/// A builder for a move call command within a programmable transaction.
+#[derive(Debug)]
+pub struct PublishBuilder<'a> {
```

Comment:
When do you create an own builder and when not? For example there is no builder for upgrade

Reply:
I created them when multiple steps were needed, some of which are optional. For instance, in this case there are two alternate ways to finish the builder. In the move call builder two of the fns are optional.